### ISSUE TITLE
Using new ObjectFactory via ServiceLoader is not working

### ISSUE THREAD

**Issue by ralphkar** [cucumber/cucumber-jvm#1691]
## Description

I have just updated from Cucumber 4.3.x to the latest 4.5.3 and discovered that our test suite now produces a warning:

    Jul 09, 2019 9:29:56 AM cucumber.runtime.java.ObjectFactoryLoader loadSelectedObjectFactory
    WARNUNG: Use deprecated reflections to load requested ObjectFactory.

We are heavily relying on Guice and use a custom ObjectFactory because we need to create a fresh (and complex) injector every time the ObjectFactory.start() method is called.

The warning states that `cucumber.api.java.ObjectFactory` is now deprecated and that `io.cucumber.core.backend.ObjectFactory` should be used instead. Looking at the changelog I noticed that ServiceLoader is now the preferred way to load the ObjectFactory.

So I changed our custom ObjectFactory to implement the new interface and created a service entry in META-INF/services so that my new ObjectFactory is being loaded.
BTW: What happened to `cucumber.properties`? Is it still in the loop. I just commented out my previous setting in there.

The observed result is now the following:

    Jul 09, 2019 9:48:04 AM cucumber.runtime.java.ObjectFactoryLoader loadObjectFactory
    WARNUNG: Expected only one instance, but found too many: [de.dfs.jarvis.web.server.GuiceObjectFactory@7eecb5b8, cucumber.runtime.java.guice.impl.GuiceFactory@30e868be]
    Jul 09, 2019 9:48:04 AM cucumber.runtime.java.ObjectFactoryLoader loadObjectFactory
    WARNUNG: More than one Cucumber ObjectFactory was found in the classpath

I have not been able to find a test or an example that demonstrates how to do this correctly. So far I have not found a way to get rid of the default `GuiceFactory`. Instead of having less capability with the `ObjectFactory` I would rather need more, because we are urgently looking for a way to use different implementations of the factory (depending on the tested feature).

## Workaround

Just continue to use the deprecated `cucumber.api.java.ObjectFactory` and `cucumber.properties` and live with the warning. But considering the speed at which your come out with new and destructive releases I assume that this workaround won't survive the next three months.

**Comment by mpkorstanje** [comment]
> What happened to cucumber.properties? Is it still in the loop. I just commented out my previous setting in there.

I suspect that is the problem. `cucumber.properties` is still used to disambiguate between multiple `ObjectFactory` candidates:

https://github.com/cucumber/cucumber-jvm/blob/master/java/src/main/java/cucumber/runtime/java/JavaBackend.java#L67

```java
    private JavaBackend(ClassFinder classFinder, TypeRegistry typeRegistry) {
        this(loadObjectFactory(classFinder, Env.INSTANCE.get(ObjectFactory.class.getName())), classFinder, typeRegistry);
    }
```

**Comment by ralphkar** [comment]
Yes, but then what is io.cucumber.core.backend.ObjectFactory and the ServiceLoader stuff good for if I cannot specify its usage explicitely.

**Comment by mpkorstanje** [comment]
Previously all `ObjectFactory` implementations would have to be located in the `cucumber.api.backend` package to be picked up by class path scanning. We are now using the [`ServiceLoader`](https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html) which allows implementations to be located in any package by adding their implementation to `io.cucumber.core.backend.ObjectFactory`.

This has two advantages:
 * We can now make `ObjectFactory` part of the public API.
 * We don't have to rely on class path scanning to discover object factories.

However when there are multiple object factories available we still need to disambiguate between them.

**Comment by ralphkar** [comment]
Understood.
But this is only an issue for internally provided factories which make Cucumber work out-of-the-box.

A customization of the `ObjectFactory` is not possible this way. The only way to disambiguate seems to be `cucumber.properties`. And the entry in there requires the factory to be of the old `cucumber.api.java.ObjectFactory` type. That way you will not be able to remove the deprecated `cucumber.api.java.ObjectFactory`.

**Comment by mpkorstanje** [comment]
Ah. I understand. Looks like the disambiguation isn't implemented for the new `ObjectFactory` yet.

https://github.com/cucumber/cucumber-jvm/blob/master/java/src/main/java/cucumber/runtime/java/ObjectFactoryLoader.java#L56

**Comment by ralphkar** [comment]
Correct.

**Comment by mpkorstanje** [comment]
I won't have time to fix this any time soon. But feel free to send a PR.

**Comment by mpkorstanje** [comment]
Fixed by #1700
